src/gtest/test_rpc_wallet.cpp

```diff
@@ -1,1258 +1,1274 @@
+    EXPECT_EQ(q->getNumberOfWorkers() , 0U);
+    vector<AsyncRPCOperationId> ids = q->getAllOperationIds();
+    EXPECT_EQ(ids.size(), 0U);
+
+    shared_ptr<AsyncRPCOperation> op1 = make_shared<AsyncRPCOperation>();
+    q->addOperation(op1);
+    EXPECT_EQ(q->getOperationCount() , 1U);
+
+    OperationStatus status = op1->getState();
+    EXPECT_EQ(status , OperationStatus::READY);
+
+    AsyncRPCOperationId id1 = op1->getId();
+    int64_t creationTime1 = op1->getCreationTime();
+
+    q->addWorker();
+    EXPECT_EQ(q->getNumberOfWorkers() , 1U);
+
+    // an AsyncRPCOperation doesn't do anything so will finish immediately
+    this_thread::sleep_for(chrono::seconds(1));
+    EXPECT_EQ(q->getOperationCount() , 0U);
+
+    // operation should be a success
+    EXPECT_FALSE(op1->isCancelled());
+    EXPECT_FALSE(op1->isExecuting());
+    EXPECT_FALSE(op1->isReady());
+    EXPECT_FALSE(op1->isFailed());
+    EXPECT_TRUE(op1->isSuccess());
+    EXPECT_TRUE(op1->getError().isNull());
+    EXPECT_FALSE(op1->getResult().isNull());
+    EXPECT_EQ(op1->getStateAsString(), "success");
+    EXPECT_NE(op1->getStateAsString(), "executing");
+
+    // Create a second operation which just sleeps
+    auto op2 = make_shared<MockSleepOperation>(2500);
+    AsyncRPCOperationId id2 = op2->getId();
+    int64_t creationTime2 = op2->getCreationTime();
+
+    // it's different from the previous operation
+    EXPECT_NE(op1.get(), op2.get());
+    EXPECT_NE(id1, id2);
+    EXPECT_NE(creationTime1, creationTime2);
+
+    // Only the first operation has been added to the queue
+    vector<AsyncRPCOperationId> v = q->getAllOperationIds();
+    set<AsyncRPCOperationId> opids(v.begin(), v.end());
+    EXPECT_EQ(opids.size() , 1U);
+    EXPECT_EQ(opids.count(id1) , 1U);
+    EXPECT_EQ(opids.count(id2) , 0U);
+    shared_ptr<AsyncRPCOperation> p1 = q->getOperationForId(id1);
+    EXPECT_EQ(p1.get(), op1.get());
+    shared_ptr<AsyncRPCOperation> p2 = q->getOperationForId(id2);
+    EXPECT_TRUE(!p2); // null ptr as not added to queue yet
+
+    // Add operation 2 and 3 to the queue
+    q->addOperation(op2);
+    auto op3 = make_shared<MockSleepOperation>(1000);
+    q->addOperation(op3);
+    this_thread::sleep_for(chrono::milliseconds(500));
+    EXPECT_TRUE(op2->isExecuting());
+    op2->cancel();  // too late, already executing
+    op3->cancel();
+    this_thread::sleep_for(chrono::milliseconds(3000));
+    EXPECT_TRUE(op2->isSuccess());
+    EXPECT_FALSE(op2->isCancelled());
+    EXPECT_TRUE(op3->isCancelled());
+
+
+    v = q->getAllOperationIds();
+    copy( v.begin(), v.end(), inserter( opids, opids.end() ) );
+    EXPECT_EQ(opids.size() , 3U);
+    EXPECT_EQ(opids.count(id1) , 1U);
+    EXPECT_EQ(opids.count(id2) , 1U);
+    EXPECT_EQ(opids.count(op3->getId()), 1U);
+    q->finishAndWait();
+}
+
+// The CountOperation will increment this global
+atomic<int64_t> gCounter(0);
+
+class CountOperation : public AsyncRPCOperation {
```

Comment:
same changes as for the MockSleepOperation class

Reply:
thanks